go/types/struct_editor.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2017 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import (
+	"sort"
+
+	"github.com/attic-labs/noms/go/d"
+)
+
+// StructEditor allows for efficient editing of Noms structs. Edits are
+// buffered to memory and can be applied via Struct(), which returns a new
+// Struct. Prior to Struct(), Get() will return the value that the resulting
+// Struct would return if it were built immediately prior to the respective call.
+// Note: The implementation biases performance towards a usage which applies
+// edits in name-order.
+type StructEditor struct {
+	s          Struct
+	edits      structEditSlice // edits may contain duplicate name values, in which case, the last edit of a given name is used
+	normalized bool
+}
+
+// NewStructEditor returns a new StructEditor starting with the fields in s.
+func NewStructEditor(s Struct) *StructEditor {
+	return &StructEditor{s, structEditSlice{}, true}
+}
+
+// Kind returns the kind of editor this is.
+func (se *StructEditor) Kind() NomsKind {
+	return StructKind
+}
+
+// Value builds a new Struct based on the edits done to the StructEditor.
+func (se *StructEditor) Value() Value {
+	return se.Struct()
+}
+
+// Struct builds a new Struct based on the edits done to the StructEditor.
+func (se *StructEditor) Struct() Struct {
+	if len(se.edits) == 0 {
+		return se.s // no edits
+	}
+
+	se.normalize()
+
+	kvsChan := make(chan chan structEntry)
+
+	go func() {
```

Comment:
We generally make the assumption that structs are "memory-sized". Can we also make the assumption that edits to structs are "memory-sized" ?

Invoking a goroutine is not cheap. This code means that if you have a large collection of structs and you edits all of them you're going to launch a goroutine for every one.

Reply:
This also means that even trivial edits to structs will end up costing one goroutine. Where before they were a stack allocation.

Reply:
You also have a channel in here. Lots of machinery to do `Set("foo", "bar")`.